src/prof_recent.c

```diff
@@ -0,0 +1,243 @@
+#define JEMALLOC_PROF_C_
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/assert.h"
+
+ssize_t		opt_prof_recent_alloc_max = 0;
+ssize_t		opt_prof_recent_dalloc_max = 0;
+
+malloc_mutex_t	prof_recent_alloc_mtx; /* Protects the fields below */
+ssize_t		prof_recent_alloc_max;
+static ssize_t prof_recent_alloc_count = 0;
+static prof_recent_alloc_t *prof_recent_alloc_head = NULL;
+static prof_recent_alloc_t *prof_recent_alloc_tail = NULL;
+
+malloc_mutex_t	prof_recent_dalloc_mtx; /* Protects the fields below */
+ssize_t		prof_recent_dalloc_max;
+
+void
+prof_recent_alloc_greet(tsdn_t *tsdn, prof_tctx_t *tctx) {
+	malloc_mutex_assert_owner(tsdn, tctx->tdata->lock);
+	malloc_mutex_assert_not_owner(tsdn, &prof_recent_alloc_mtx);
+
+	/*
+	 * Increment ref_count to hold the tctx so that it won't be gone even
+	 * after tctx->tdata->lock is released.  This acts as a "placeholder";
+	 * the real recording of the allocation requires a lock on
+	 * prof_recent_alloc_mtx and is done in prof_recent_alloc_slow (when
+	 * tctx->tdata->lock has been released).
+	 */
+	++tctx->ref_count;
+}
+
+static void
+decrement_ref_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	/*
+	 * tdata->lock and prof_recent_alloc_mtx should not be held
+	 * simultaneously.
+	 */
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->ref_count > 0);
+	--tctx->ref_count;
+	if (prof_tctx_should_destroy(tsd_tsdn(tsd), tctx)) {
+		prof_tctx_destroy(tsd, tctx);
+	} else {
+		malloc_mutex_unlock(tsd_tsdn(tsd), tctx->tdata->lock);
+	}
+}
+
+bool
+prof_recent_alloc(tsd_t *tsd, const void *ptr, size_t usize,
+    prof_tctx_t *tctx) {
+	/*
+	 * tdata->lock and prof_recent_alloc_mtx should not be held
+	 * simultaneously.
+	 */
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+
+	prof_recent_alloc_t *node;
+	prof_tctx_t *head_tctx;
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (prof_recent_alloc_max == 0) {
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		/*
+		 * The increment of ref_count in prof_recent_alloc_fast was a
+		 * mistake - fix it.
+		 */
+		decrement_ref_count(tsd, tctx);
+		return false;
+	}
+
+	if (prof_recent_alloc_max != -1 &&
+	    prof_recent_alloc_count == prof_recent_alloc_max) {
+		/* If the upper limit is reached, simply reuse the head node. */
+		assert(prof_recent_alloc_head != NULL);
+		assert(prof_recent_alloc_tail != NULL);
+		assert(prof_recent_alloc_count > 0);
+		node = prof_recent_alloc_head;
+		prof_recent_alloc_head = node->next;
+		if (prof_recent_alloc_head == NULL) {
+			prof_recent_alloc_tail = NULL;
+		}
+		--prof_recent_alloc_count;
```

Comment:
This feels a bit strange; should be better to not adjust the count in this branch, and simply bump it in the new alloc node case -- so that the alloc_count is not adjusted at all in the common case (where we don't change alloc_max, and keep reusing nodes).